File: connectors/java/src/test/java/com/apiglimpse/RedactionTest.java
package com.apiglimpse;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class RedactionTest {

  private final ObjectMapper mapper = Envelope.mapper();

  @Test
  void envelopeVersionIsOne() {
    assertEquals(1, Redaction.ENVELOPE_VERSION);
  }

  @Test
  void sensitiveHeadersSet() {
    assertTrue(Redaction.SENSITIVE_HEADER_NAMES.contains("authorization"));
    assertTrue(Redaction.SENSITIVE_HEADER_NAMES.contains("cookie"));
    assertTrue(Redaction.SENSITIVE_HEADER_NAMES.contains("set-cookie"));
    assertTrue(Redaction.SENSITIVE_HEADER_NAMES.contains("x-api-key"));
    assertTrue(Redaction.SENSITIVE_HEADER_NAMES.contains("x-auth-token"));
    assertTrue(Redaction.SENSITIVE_HEADER_NAMES.contains("proxy-authorization"));
  }

  @Test
  void truncateString() {
    assertEquals("short", Redaction.truncateString("short"));
    String longStr = "x".repeat(100);
    String out = Redaction.truncateString(longStr, 64);
    assertTrue(out.endsWith("…"));
    assertEquals(65, out.length());
  }

  @Test
  void redactHeaders() {
    Map<String, String> headers = new LinkedHashMap<>();
    headers.put("Authorization", "Bearer secret-token");
    headers.put("Content-Type", "application/json");
    headers.put("X-Request-Id", "req-1");
    headers.put("Cookie", "session=abc");

    Map<String, String> out = Redaction.redactHeaders(headers);
    assertEquals("[REDACTED]", out.get("authorization"));
    assertEquals("[REDACTED]", out.get("cookie"));
    assertEquals("application/json", out.get("content-type"));
    assertEquals("req-1", out.get("x-request-id"));
  }

  @Test
  void redactValuePatterns() {
    assertEquals("Bearer [REDACTED]", Redaction.redactValue("Bearer abc.def"));
    assertEquals(
        "[REDACTED_JWT]",
        Redaction.redactValue("eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.payload.signature"));
    assertEquals("[REDACTED_SSN]", Redaction.redactValue("ssn 123-45-6789 here"));
    assertEquals("[REDACTED_CARD]", Redaction.redactValue("4111 1111 1111 1111"));
    assertEquals("hello", Redaction.redactValue("hello"));
  }

  @Test
  void shapeBodyPrimitives() {
    assertEquals("null", Redaction.shapeBody(null).get("type"));
    assertEquals("hi", Redaction.shapeBody("hi").get("sample"));
    assertEquals(true, Redaction.shapeBody(true).get("sample"));
    assertEquals("integer", Redaction.shapeBody(42).get("type"));
    assertEquals("number", Redaction.shapeBody(3.14).get("type"));
  }

  @Test
  void shapeBodySecretKeys() {
    Map<String, Object> body = new LinkedHashMap<>();
    body.put("email", "user@example.com");
    body.put("password", "hunter2");
    body.put("api_token", "tok_live");
    body.put("ssn", "123-45-6789");
    body.put("cvv", "123");
    Map<String, Object> profile = new LinkedHashMap<>();
    profile.put("name", "Ada");
    profile.put("age", 36);
    body.put("profile", profile);

    Map<String, Object> shaped = Redaction.shapeBody(body);
    @SuppressWarnings("unchecked")
    Map<String, Object> props = (Map<String, Object>) shaped.get("properties");
    assertEquals("[REDACTED]", ((Map<?, ?>) props.get("password")).get("sample"));
    assertEquals("[REDACTED]", ((Map<?, ?>) props.get("api_token")).get("sample"));
    assertEquals("[REDACTED]", ((Map<?, ?>) props.get("ssn")).get("sample"));
    assertEquals("[REDACTED]", ((Map<?, ?>) props.get("cvv")).get("sample"));
    assertEquals("user@example.com", ((Map<?, ?>) props.get("email")).get("sample"));
    assertEquals(false, shaped.get("truncatedKeys"));
  }

  @Test
  void shapeBodyCaps() {
    Map<String, Object> deep = Map.of("a", Map.of("b", Map.of("c", Map.of("d", Map.of("e", "too deep")))));
    Map<String, Object> shaped = Redaction.shapeBody(deep);
    @SuppressWarnings("unchecked")
    Map<String, Object> a = (Map<String, Object>) ((Map<?, ?>) shaped.get("properties")).get("a");
    @SuppressWarnings("unchecked")
    Map<String, Object> b = (Map<String, Object>) ((Map<?, ?>) a.get("properties")).get("b");
    @SuppressWarnings("unchecked")
    Map<String, Object> c = (Map<String, Object>) ((Map<?, ?>) b.get("properties")).get("c");
    @SuppressWarnings("unchecked")
    Map<String, Object> d = (Map<String, Object>) ((Map<?, ?>) c.get("properties")).get("d");
    assertEquals("truncated", d.get("type"));

    List<Integer> many = new ArrayList<>();
    for (int i = 0; i < 10; i++) {
      many.add(i);
    }
    Map<String, Object> arr = Redaction.shapeBody(many);
    assertEquals(10, arr.get("length"));
    assertEquals(5, ((List<?>) arr.get("items")).size());

    Map<String, Object> keys = new LinkedHashMap<>();
    for (int i = 0; i < 50; i++) {
      keys.put("k" + i, i);
    }
    Map<String, Object> obj = Redaction.shapeBody(keys);
    assertEquals(40, ((Map<?, ?>) obj.get("properties")).size());
    assertEquals(true, obj.get("truncatedKeys"));
  }

  @Test
  void shapeBodyMatchesFixture() throws Exception {
    Map<String, Object> fixture = loadFixture("sample-shaped.json");
    Map<String, Object> reqBody = new LinkedHashMap<>();
    reqBody.put("email", "user@example.com");
    reqBody.put("password", "s3cret");
    Map<String, Object> profile = new LinkedHashMap<>();
    profile.put("name", "Ada");
    profile.put("age", 36);
    reqBody.put("profile", profile);

    Map<String, Object> resBody = new LinkedHashMap<>();
    resBody.put("id", "usr_01");
    resBody.put("email", "user@example.com");
    resBody.put("token", "tok_live");

    @SuppressWarnings("unchecked")
    Map<String, Object> wantReq =
        (Map<String, Object>) ((Map<?, ?>) fixture.get("request")).get("bodyShape");
    @SuppressWarnings("unchecked")
    Map<String, Object> wantRes =
        (Map<String, Object>) ((Map<?, ?>) fixture.get("response")).get("bodyShape");

    assertJsonEqual(wantReq, Redaction.shapeBody(reqBody));
    assertJsonEqual(wantRes, Redaction.shapeBody(resBody));
  }

  @SuppressWarnings("unchecked")
  private Map<String, Object> loadFixture(String name) throws Exception {
    try (InputStream in = getClass().getResourceAsStream("/fixtures/" + name)) {
      assertTrue(in != null, "missing fixture " + name);
      return mapper.readValue(in, Map.class);
    }
  }

  private void assertJsonEqual(Object want, Object got) throws Exception {
    Object wn = mapper.readValue(mapper.writeValueAsBytes(want), Object.class);
    Object gn = mapper.readValue(mapper.writeValueAsBytes(got), Object.class);
    assertEquals(wn, gn);
  }
}

File: connectors/go/apiglimpse/redaction_test.go
package apiglimpse

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

func TestEnvelopeVersion(t *testing.T) {
	if EnvelopeVersion != 1 {
		t.Fatalf("EnvelopeVersion = %d, want 1", EnvelopeVersion)
	}
}

func TestRedactHeaders(t *testing.T) {
	in := map[string]string{
		"Content-Type":  "application/json",
		"Authorization": "Bearer secret-token",
		"X-Request-Id":  "req-abc-123",
		"Cookie":        "sid=1",
	}
	out := RedactHeaders(in)
	if out["authorization"] != "[REDACTED]" {
		t.Fatalf("authorization: %q", out["authorization"])
	}
	if out["cookie"] != "[REDACTED]" {
		t.Fatalf("cookie: %q", out["cookie"])
	}
	if out["content-type"] != "application/json" {
		t.Fatalf("content-type: %q", out["content-type"])
	}
	if out["x-request-id"] != "req-abc-123" {
		t.Fatalf("x-request-id: %q", out["x-request-id"])
	}
}

func TestRedactValue(t *testing.T) {
	cases := []struct {
		in, want string
	}{
		{"Bearer abc", "Bearer [REDACTED]"},
		{"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.demo.signature", "[REDACTED_JWT]"},
		{"ssn 123-45-6789 here", "[REDACTED_SSN]"},
		{"short", "short"},
	}
	for _, tc := range cases {
		if got := RedactValue(tc.in); got != tc.want {
			t.Fatalf("RedactValue(%q)=%q want %q", tc.in, got, tc.want)
		}
	}
}

func TestShapeBodyMatchesFixture(t *testing.T) {
	fixture := loadJSON(t, "sample-shaped.json")

	reqBody := map[string]any{
		"email":    "user@example.com",
		"password": "s3cret",
		"profile": map[string]any{
			"name": "Ada",
			"age":  36,
		},
	}
	resBody := map[string]any{
		"id":    "usr_01",
		"email": "user@example.com",
		"token": "tok_live",
	}

	reqShape := ShapeBody(reqBody, 0)
	resShape := ShapeBody(resBody, 0)

	wantReq := fixture["request"].(map[string]any)["bodyShape"]
	wantRes := fixture["response"].(map[string]any)["bodyShape"]

	assertJSONEqual(t, wantReq, reqShape)
	assertJSONEqual(t, wantRes, resShape)
}

func TestCreateSampleMatchesFixture(t *testing.T) {
	want := loadJSON(t, "sample-shaped.json")

	sample := CreateSample(SampleInput{
		Method:     "POST",
		Path:       "/api/users",
		StatusCode: 201,
		LatencyMs:  42,
		RequestHeaders: map[string]string{
			"content-type":  "application/json",
			"authorization": "Bearer secret",
			"x-request-id":  "req-abc-123",
		},
		ResponseHeaders: map[string]string{
			"content-type": "application/json",
			"set-cookie":   "sid=abc",
		},
		RequestHeaderNames:  []string{"content-type", "authorization", "x-request-id"},
		ResponseHeaderNames: []string{"content-type", "set-cookie"},
		RequestBody: map[string]any{
			"email":    "user@example.com",
			"password": "s3cret",
			"profile": map[string]any{
				"name": "Ada",
				"age":  36,
			},
		},
		ResponseBody: map[string]any{
			"id":    "usr_01",
			"email": "user@example.com",
			"token": "tok_live",
		},
		HasRequestBody:  true,
		HasResponseBody: true,
		AuthObserved:    "bearer",
		Timestamp:       "2026-01-15T12:00:00.000Z",
	})

	gotBytes, err := json.Marshal(sample)
	if err != nil {
		t.Fatal(err)
	}
	var got map[string]any
	if err := json.Unmarshal(gotBytes, &got); err != nil {
		t.Fatal(err)
	}
	assertJSONEqual(t, want, got)
}

func TestCreateEnvelopeMinimalFixture(t *testing.T) {
	want := loadJSON(t, "envelope-v1-minimal.json")
	env := Envelope{
		Version: EnvelopeVersion,
		APIKey:  "ask_minimal",
		Samples: []Sample{},
		SentAt:  "2026-01-15T12:00:00.000Z",
	}
	gotBytes, err := json.Marshal(env)
	if err != nil {
		t.Fatal(err)
	}
	var got map[string]any
	if err := json.Unmarshal(gotBytes, &got); err != nil {
		t.Fatal(err)
	}
	assertJSONEqual(t, want, got)
}

func TestCreateEnvelopeSampleFixture(t *testing.T) {
	want := loadJSON(t, "envelope-v1-sample.json")
	sample := CreateSample(SampleInput{
		Method:     "POST",
		Path:       "/api/users",
		StatusCode: 201,
		LatencyMs:  42,
		RequestHeaders: map[string]string{
			"content-type":  "application/json",
			"authorization": "Bearer secret",
			"x-request-id":  "req-abc-123",
		},
		ResponseHeaders: map[string]string{
			"content-type": "application/json",
			"set-cookie":   "sid=abc",
		},
		RequestHeaderNames:  []string{"content-type", "authorization", "x-request-id"},
		ResponseHeaderNames: []string{"content-type", "set-cookie"},
		RequestBody: map[string]any{
			"email":    "user@example.com",
			"password": "s3cret",
			"profile": map[string]any{
				"name": "Ada",
				"age":  36,
			},
		},
		ResponseBody: map[string]any{
			"id":    "usr_01",
			"email": "user@example.com",
			"token": "tok_live",
		},
		HasRequestBody:  true,
		HasResponseBody: true,
		AuthObserved:    "bearer",
		Timestamp:       "2026-01-15T12:00:00.000Z",
	})
	env := Envelope{
		Version: EnvelopeVersion,
		APIKey:  "ask_test_key_fixture",
		Samples: []Sample{sample},
		SentAt:  "2026-01-15T12:00:01.000Z",
	}
	gotBytes, err := json.Marshal(env)
	if err != nil {
		t.Fatal(err)
	}
	var got map[string]any
	if err := json.Unmarshal(gotBytes, &got); err != nil {
		t.Fatal(err)
	}
	assertJSONEqual(t, want, got)
}

func TestValidateEnvelope(t *testing.T) {
	ok, _ := ValidateEnvelope(&Envelope{Version: 1, Samples: []Sample{}})
	if !ok {
		t.Fatal("expected ok")
	}
	ok, _ = ValidateEnvelope(&Envelope{Version: 2, Samples: []Sample{}})
	if ok {
		t.Fatal("expected reject v2")
	}
}

func TestObserveAuth(t *testing.T) {
	if got := ObserveAuth(map[string]string{"authorization": "Bearer x"}); got != "bearer" {
		t.Fatalf("got %q", got)
	}
	if got := ObserveAuth(map[string]string{"cookie": "a=b"}); got != "cookie" {
		t.Fatalf("got %q", got)
	}
	if got := ObserveAuth(map[string]string{}); got != "none" {
		t.Fatalf("got %q", got)
	}
}

func loadJSON(t *testing.T, name string) map[string]any {
	t.Helper()
	b, err := os.ReadFile(filepath.Join("testdata", name))
	if err != nil {
		t.Fatal(err)
	}
	var out map[string]any
	if err := json.Unmarshal(b, &out); err != nil {
		t.Fatal(err)
	}
	return out
}

func assertJSONEqual(t *testing.T, want, got any) {
	t.Helper()
	wn := normalizeJSON(want)
	gn := normalizeJSON(got)
	if !reflect.DeepEqual(wn, gn) {
		wb, _ := json.MarshalIndent(wn, "", "  ")
		gb, _ := json.MarshalIndent(gn, "", "  ")
		t.Fatalf("mismatch\nwant:\n%s\ngot:\n%s", wb, gb)
	}
}

// normalizeJSON re-encodes via encoding/json so number kinds align (float64).
func normalizeJSON(v any) any {
	b, err := json.Marshal(v)
	if err != nil {
		return v
	}
	var out any
	if err := json.Unmarshal(b, &out); err != nil {
		return v
	}
	return out
}
